daemon/access.go

```diff
@@ -151,3 +154,91 @@ func (ac snapAccess) CheckAccess(d *Daemon, r *http.Request, ucred *ucrednet, us
 	// FIXME: should snapctl access be allowed on the main socket?
 	return Forbidden("access denied")
 }
+
+var (
+	cgroupSnapNameFromPid = cgroup.SnapNameFromPid
+	requireThemeApiAccess = requireThemeApiAccessImpl
+)
+
+func requireThemeApiAccessImpl(d *Daemon, ucred *ucrednet) *apiError {
+	if ucred == nil {
+		return Forbidden("access denied")
+	}
+
+	switch ucred.Socket {
+	case dirs.SnapdSocket:
+		// Allow access on main snapd.socket
+		return nil
+
+	case dirs.SnapSocket:
+		// Handled below
+	default:
+		return Forbidden("access denied")
+	}
+
+	// Access on snapd-snap.socket requires a connected
+	// snapd-themes-control plug.
+	snapName, err := cgroupSnapNameFromPid(int(ucred.Pid))
+	if err != nil {
+		return Forbidden("could not determine snap name for pid: %s", err)
+	}
+
+	st := d.state
+	st.Lock()
+	defer st.Unlock()
+	conns, err := ifacestate.ConnectionStates(st)
+	if err != nil {
+		return Forbidden("internal error: cannot get connections: %s", err)
+	}
+	for refStr, connState := range conns {
+		if connState.Undesired || connState.HotplugGone || connState.Interface != "snapd-themes-control" {
+			continue
+		}
+		connRef, err := interfaces.ParseConnRef(refStr)
+		if err != nil {
+			return Forbidden("internal error: %s", err)
+		}
+		if connRef.PlugRef.Snap == snapName {
+			return nil
+		}
+	}
+	return Forbidden("access denied")
+}
+
+// themesOpenAccess behaves like openAccess, but allows requests from
+// snapd-snap.socket for snaps that plug snapd-themes-control.
+type themesOpenAccess struct{}
+
+func (ac themesOpenAccess) CheckAccess(d *Daemon, r *http.Request, ucred *ucrednet, user *auth.UserState) *apiError {
+	return requireThemeApiAccess(d, ucred)
+}
+
+// themesAuthenticatedAccess behaves like authenticatedAccess, but
+// allows requests from snapd-snap.socket for snaps that plug
+// snapd-themes-control.
+type themesAuthenticatedAccess struct {
+	Polkit string
+}
+
+func (ac themesAuthenticatedAccess) CheckAccess(d *Daemon, r *http.Request, ucred *ucrednet, user *auth.UserState) *apiError {
+	if rspe := requireThemeApiAccess(d, ucred); rspe != nil {
```

Comment:
shouldn't the check be rspe == nil?  I don't understand what this is doing concretely

Reply:
`requireThemeApiAccess` returns a non-nil `*apiError` if the client connected on `snapd-snap.socket` and is not a snap with `snapd-themes-control` plugged. If it fails that test, then we want to deny access.

But that isn't sufficient to grant access: we also need to check that the client has administrative access to the system, similar to the `authenticatedAccess` checker.

Reply:
I'm still confused by this, we will have to talk about it when I'm back from holidays (I'm off next week)

Reply:
The intent is that a polkit dialog would be shown when the API is invoked by the snapd-desktop-integration helper service (since you need administrative access to install snaps). If we returned a nil here, the polkit code path would never be reached.